diesel_cli/src/main.rs

```diff
@@ -110,62 +126,120 @@ fn run_setup_command(matches: &ArgMatches) {
                         .map_err(handle_error).unwrap()
                        );
 
-    if PgConnection::establish(&database_url(matches)).is_err() {
-        create_database(database_url(matches)).unwrap_or_else(handle_error);
-    }
+    setup_database(matches).unwrap_or_else(handle_error);
+}
+
+fn run_database_command(matches: &ArgMatches) {
+    match matches.subcommand() {
+        ("setup", Some(args)) => setup_database(args).unwrap_or_else(handle_error),
+        ("reset", Some(args)) => reset_database(args).unwrap_or_else(handle_error),
+        _ => unreachable!("The cli parser should prevent reaching here"),
+    };
+}
+
+fn reset_database(args: &ArgMatches) -> Result<(), DatabaseError> {
+    let (database, postgres_url) = split_pg_connection_string(&database_url(args));
+    let connection = try!(PgConnection::establish(&postgres_url));
+    try!(drop_database(&connection, &database));
```

Comment:
I don't think `drop_database` should take a connection. The process of doing this will vary too much by backend.